cpp/src/utilities/type_dispatcher.hpp

```diff
@@ -107,14 +114,89 @@ decltype(auto) type_dispatcher(gdf_dtype dtype,
     case GDF_DATE64:    { return f.template operator()< date64 >(std::forward<Ts>(args)...); }
     case GDF_TIMESTAMP: { return f.template operator()< timestamp >(std::forward<Ts>(args)...); }
     case GDF_CATEGORY:  { return f.template operator()< category >(std::forward<Ts>(args)...); }
-    default:            { assert(false && "type_dispatcher: invalid gdf_type"); }
+    default: { assert(false && "type_dispatcher: invalid gdf_type"); }
```

Comment:
Can you do something other than assert to make your death tests work without changing the value of `NDEBUG`?

Reply:
Thanks to @eyalroz who found that `assert` is calling an `__assertfail` intrinsic in device code, we could just call that directly. In that way, it will always be called regardless of the state of `NDEBUG`. 

See https://github.com/rapidsai/cudf/issues/527#issuecomment-463295623

Reply:
@jrhemstad : But note that `__assertfail()` may be compiler-dependent, so I'm not sure calling it directly is necessarily a great idea.

Reply:
> @jrhemstad : But note that `__assertfail()` may be compiler-dependent, so I'm not sure calling it directly is necessarily a great idea.

It's certainly dependent on the `nvcc` compiler as it's a device intrinsic. Apart from that, I don't expect it to depend on any particular version of `nvcc`.

So as long as we wrap it in a `#ifdef _CUDACC__` then we'll be good to go. I think for the host-compiled version, we can just throw an exception.